core/src/main/java/com/exadel/aem/toolkit/bundle/lists/datasource/ChildResourcesDatasource.java

```diff
@@ -0,0 +1,197 @@
+        request.setAttribute(DataSource.class.getName(), dataSource);
+    }
+
+    /**
+     * Retrieves the list of item resources, which are either lists themselves, or folders
+     * that may contain lists inside
+     *
+     * @param resolver An instance of ResourceResolver
+     * @param parent   {@code Resource} instance used as the source of markup
+     * @return a list of {@link Resource}s
+     */
+    private List<Resource> getValidChildren(ResourceResolver resolver, Resource parent) {
+        return getChildrenStream(parent)
+            .filter(resource -> isTemplate(resolver, resource) || isFolder(resource))
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Checks whether the resource is a list page
+     *
+     * @param resolver An instance of ResourceResolver
+     * @param resource {@code Resource} instance used as the source of markup
+     * @return True or false
+     */
+    private static boolean isTemplate(ResourceResolver resolver, Resource resource) {
+        Resource childParameters = resolver.getResource(resource.getPath() + PATH_TO_JCR_CONTENT);
+        if (childParameters != null) {
+            String template = childParameters.getValueMap().get(NameConstants.NN_TEMPLATE, String.class);
+            return template != null && template.equals(LIST_TEMPLATE_NAME);
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether the resource contains resources, which are either pages, or folders that may contain lists inside
+     *
+     * @param resource {@code Resource} instance used as the source of markup
+     * @return True or false
+     */
+    private static boolean isFolder(Resource resource) {
+        return getChildrenStream(resource).anyMatch(item -> {
+            String primaryType = item.getValueMap().get(JcrConstants.JCR_PRIMARYTYPE, String.class);
+            return primaryType != null && (primaryType.equals(NameConstants.NT_PAGE) || primaryType.equals(JcrConstants.NT_FOLDER));
+        });
+    }
+
+    private static Stream<Resource> getChildrenStream(Resource resource) {
+        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(resource.listChildren(), Spliterator.ORDERED), false);
+    }
+
+    /**
+     * Gets {@code request} sling script helper
+     *
+     * @param request {@code ServletRequest} instance
+     * @return {@code SlingScriptHelper} instance
+     */
+    private static SlingScriptHelper getScriptHelper(ServletRequest request) {
+        SlingBindings bindings = (SlingBindings) request.getAttribute(SlingBindings.class.getName());
+        return bindings.getSling();
+    }
+
+    public class CustomAbstractDataSource extends AbstractDataSource {
+        private Iterator<Resource> resources;
+        private Integer offset;
+        private Integer limit;
+        private String itemResourceType;
+
+        private CustomAbstractDataSource(Iterator<Resource> resources, Integer offset, Integer limit, String itemResourceType) {
```

Comment:
Please revise the usage of boxed primitives here as well

Reply:
Done.